File: sbe-tool/src/main/java/uk/co/real_logic/sbe/json/JsonPrinter.java
/*
 * Copyright 2013-2020 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.json;

import org.agrona.concurrent.UnsafeBuffer;
import uk.co.real_logic.sbe.ir.Ir;
import uk.co.real_logic.sbe.ir.Token;
import uk.co.real_logic.sbe.otf.OtfHeaderDecoder;
import uk.co.real_logic.sbe.otf.OtfMessageDecoder;

import java.nio.ByteBuffer;
import java.util.List;

/**
 * Pretty Print JSON based upon the given Ir.
 */
public class JsonPrinter
{
    private final OtfHeaderDecoder headerDecoder;
    private final Ir ir;

    public JsonPrinter(final Ir ir)
    {
        this.ir = ir;
        headerDecoder = new OtfHeaderDecoder(ir.headerStructure());
    }

    public void print(final ByteBuffer encodedMessage, final StringBuilder output)
    {
        final UnsafeBuffer buffer = new UnsafeBuffer(encodedMessage);
        print(output, buffer, 0);
    }

    public void print(final StringBuilder output, final UnsafeBuffer buffer, final int bufferOffset)
    {
        final int blockLength = headerDecoder.getBlockLength(buffer, bufferOffset);
        final int templateId = headerDecoder.getTemplateId(buffer, bufferOffset);
        final int schemaId = headerDecoder.getSchemaId(buffer, bufferOffset);
        final int actingVersion = headerDecoder.getSchemaVersion(buffer, bufferOffset);

        validateId(schemaId);

        final int messageOffset = bufferOffset + headerDecoder.encodedLength();
        final List<Token> msgTokens = ir.getMessage(templateId);

        OtfMessageDecoder.decode(
            buffer,
            messageOffset,
            actingVersion,
            blockLength,
            msgTokens,
            new JsonTokenListener(output));
    }

    private void validateId(final int schemaId)
    {
        if (schemaId != ir.id())
        {
            throw new IllegalArgumentException("Required schema id " + ir.id() + " but was " + schemaId);
        }
    }

    public String print(final ByteBuffer encodedMessage)
    {
        final StringBuilder sb = new StringBuilder();
        print(encodedMessage, sb);

        return sb.toString();
    }
}

File: sbe-tool/src/main/golang/json/json_printer.go
package json

import (
	"fmt"
	"strings"

	"github.com/real-logic/simple-binary-encoding/otf"
)

// JsonPrinter pretty prints JSON based upon the given Ir.
type JsonPrinter struct {
	headerDecoder *otf.OftHeaderDecoder
	ir            *otf.IrDecoder
}

// NewJsonPrinter creates a new JsonPrinter for a given message Ir.
func NewJsonPrinter(ir *otf.IrDecoder) (*JsonPrinter, error) {
	d, err := otf.NewOtfHeaderDecoder(ir.Header())
	return &JsonPrinter{
		headerDecoder: d,
		ir:            ir,
	}, err
}

// Print the encoded message to the output.
func (printer *JsonPrinter) Print(buffer []byte, output *strings.Builder) error {
	blockLength, err := printer.headerDecoder.BlockLength(buffer)
	if err != nil {
		return fmt.Errorf("Error getting blockLength: %s", err)
	}
	templateId, err := printer.headerDecoder.TemplateId(buffer)
	if err != nil {
		return fmt.Errorf("Error getting templateId: %s", err)
	}
	schemaId, err := printer.headerDecoder.SchemaId(buffer)
	if err != nil {
		return fmt.Errorf("Error getting schemaId: %s", err)
	}
	actingVersion, err := printer.headerDecoder.SchemaVersion(buffer)
	if err != nil {
		return fmt.Errorf("Error getting schemaVersion: %s", err)
	}
	err = printer.validateId(schemaId)
	if err != nil {
		return err
	}

	messageOffset := printer.headerDecoder.EncodedLength()
	msgTokens := printer.ir.MessageByID(int32(templateId))

	otf.Decode(
		buffer[messageOffset:],
		actingVersion,
		blockLength,
		msgTokens,
		newJsonTokenListener(output),
	)

	return nil
}

func (printer *JsonPrinter) validateId(schemaId uint64) error {
	if schemaId != uint64(printer.ir.Id()) {
		return fmt.Errorf("Required schema id %d but was %d", printer.ir.Id(), schemaId)
	}
	return nil
}

// PrintJson the encoded message to a String.
func (printer *JsonPrinter) PrintJson(buffer []byte) (string, error) {
	sb := strings.Builder{}
	if err := printer.Print(buffer, &sb); err != nil {
		return "", err
	}
	return sb.String(), nil
}
